### DISCUSSION TITLE
Feature Request: Parallelism per-executor

### DISCUSSION THREAD

**Discussion by czycha** [nrwl/nx#23275] [Feature Requests]
- [ ] I'd be willing to implement this feature ([contributing guide](https://github.com/nrwl/nx/blob/master/CONTRIBUTING.md))

## Description
I would like to be able to control parallelism as a default at the executor level. This could either be through the nx.json file or perhaps configuration on the executor configuration itself for its maximum level of parallelism.

## Motivation
This would be beneficial because some tasks cannot work in parallel. For example, I've written an [LFS](https://git-lfs.com/) download task, but the LFS commands seem to fail if run in parallel with each other (even though they aren't touching the same files). This is an issue because we want it to be a pre-requisite for build tasks which we **do** want in parallel. So the ideal flow would be to run `nx run-many -t=build` and it runs all the LFS commands serially (`--parallel=1`) and then the build commands with the default parallelism (`--parallel=3`).

As a workaround, I've had to remove the LFS tasks as dependencies of the build ones (even though they are) and wrote up documentation on this in our README that if someone wants to build many projects at once, they have to first LFS download with `--parallel=1` and then run their builds.

### Configuring
The nx.json is perhaps the easiest solve since it already accepts defaults. It just needs to have and understand a `parallel` field.

However, I think the more scalable solution would be having it at the executor definition level and not on the nx.json. If I, for example, decided to publish my LFS executor I know that everyone should have this constraint of max parallelism. So something like `"maxParallel": 1` in the executor.json file. By default, there is no max. 

I don't think that we have to have one or the other: I think the executor.json fixes the root issue in all cases, but the nx.json could be convenient for other reasons.

### Running

This is where it gets tricky. In my specific case, I'm trying to keep multiple tasks with the same executor from running in parallel. So it actually wouldn't matter too much if there are many tasks being run in parallel, just that the LFS executors have a limit of 1 per pool of tasks. But I'm not sure that this is a friendly and expected notion of what a parallel flag would do.

So my guess is to have an approach where, for any given possible set of tasks that can be run, it runs the minimum parallelism and widens/narrows depending on the next possible batch of tasks. Since we want to do it dependency order, this is potentially some major tetris work.

**Comment by FrozenPandaz** [comment]
I've converted this to a discussion so that it can be discussed more easily.

**Comment by czycha** [comment]
It seems this has been done in https://github.com/nrwl/nx/pull/26820! Awesome work @xiongemi
